File: Kolokwium 13.11.2017/find_longest.m
% Date: 13.11.2017
% Author: Micha³ Kleszczyñski

function [ data_start, data_end ] = find_longest( data )
    data_avg = mean( data );

    max_start_idx = 1;
    max_end_idx = 1;
    max_length = 1;

    current_start_idx = 1;
    current_end_idx = 1;
    current_length = 1;
    
    for i = 1:length( data )
    	% If value if bigger than average
    	if data( i ) >= data_avg
    		% If current data is longer than the maximum
     		if current_length > max_length
     			max_start_idx=current_start_idx;
                max_end_idx=current_end_idx;
     			max_length=current_length;
            end
            
            % Set current position to the next value
            if i < length( data )
                current_start_idx=i+1;
                current_end_idx=i+1;
                current_length=1;
            end
            
    	else 	% In every other case
    		current_end_idx=i;
    		current_length=current_length+1;
    	end
    end
    
    data_start = max_start_idx;
    data_end = max_end_idx;
end

File: Kolokwium 13.11.2017/find_longest.cpp

// [ 7 3 3 4 4 7 8 7 3 4 ]

std::vector<double> v;

std::pair<size_t, size_t> find_longest(std::vector<double> v)
{
	double avg = std::avg(v.begin(), v.end());

	size_t max_start_idx = 0;
	size_t max_end_idx = 0;
	size_t max_length = 1;

	size_t current_start_id = 0;
	size_t current_end_idx = 0;
	size_t current_length = 1;

	for(int i = 0; i < v.length(); i++) {
		// Value bigger than average
		if ( v[i] >= avg ) {			
			if ( current_length > max_length ) {
				max_start_idx 		= current_start_id;
				max_end_idx 		= current_end_idx;
				max_length 			= current_length;
			}

			if ( i < v.length() ) {
				current_start_id 	= i + 1;
				current_end_idx 	= i + 1;
				current_length 		= 1;
			}
		} else {				// In other case
			current_end_idx = i;
			current_length++;
		}
	}

	return std::make_pair( max_start_idx, max_end_idx );
}